android-runtime/runtime/src/main/java/io/sdkgen/runtime/SdkgenHttpClient.kt

```diff
@@ -0,0 +1,335 @@
+
+        if (region.isNotEmpty()) {
+            bcp47Tag.append(sep).append(region)
+        }
+
+        if (variant.isNotEmpty()) {
+            bcp47Tag.append(sep).append(variant)
+        }
+
+        return bcp47Tag.toString()
+    }
+
+    @SuppressLint("HardwareIds")
+    private fun makeDeviceObj(): JsonObject {
+        return JsonObject().apply {
+            val pref = applicationContext.getSharedPreferences("api", Context.MODE_PRIVATE)
+
+            if (pref.contains("deviceId")) {
+                addProperty("id", pref.getString("deviceId", null))
+            }
+
+            addProperty("fingerprint", Settings.Secure.getString(applicationContext.contentResolver, Settings.Secure.ANDROID_ID))
+            addProperty("language", language())
+            add("platform", JsonObject().apply {
+                addProperty("version", Build.VERSION.RELEASE)
+                addProperty("sdkVersion", Build.VERSION.SDK_INT)
+                addProperty("brand", Build.BRAND)
+                addProperty("model", Build.MODEL)
+                add("screen", JsonObject().apply {
+                    val manager = applicationContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+                    val display = manager.defaultDisplay
+                    val size = Point()
+                    display.getSize(size)
+                    addProperty("width", size.x)
+                    addProperty("height", size.y)
+                })
+            })
+
+            addProperty("timezone", Calendar.getInstance().timeZone.displayName)
+            addProperty("type", "android")
+
+            try {
+                addProperty("version", applicationContext.packageManager.getPackageInfo(applicationContext.packageName, 0).versionName)
+            } catch (e: Exception) { }
+        }
+    }
+
+    @SuppressLint("HardwareIds")
+    suspend fun makeRequest(
+        functionName: String,
+        bodyArgs: JsonObject?,
+        timeoutMillis: Long? = null
+    ): InternalResponse = suspendCoroutine { continuation ->
+        try {
+            val body = JsonObject().apply {
+                addProperty("version", 3)
+                addProperty("requestId", callId())
+                addProperty("name", functionName)
+                add("args", bodyArgs ?: JsonObject())
+                add("deviceInfo", makeDeviceObj())
+            }
+
+            Log.d("tag body", body.toString())
+
+            val request = Request.Builder()
+                .url("$baseUrl${if (baseUrl.last() == '/') "" else "/"}$functionName")
+                .post(body.toString().toRequestBody("application/json; charset=utf-8".toMediaType()))
+                .build()
+
+            val call = httpClient.newCall(request)
+            call.timeout().timeout(timeoutMillis ?: defaultTimeoutMillis, TimeUnit.MILLISECONDS)
+
+            try {
+                val response = call.execute()
+
+                if (response.code in 500..599) {
```

Comment:
Error 500 is returned by sdkgen in case of fatal errors. for those the error message come in the body and should be parsed. Change to 501 here.